flink-ml-lib/src/main/java/org/apache/flink/ml/feature/onehotencoder/OneHotEncoder.java

```diff
@@ -0,0 +1,146 @@
+                tEnv.toDataStream(inputs[0])
+                        .flatMap(new ExtractInputColsValueFunction(inputCols))
+                        .keyBy(x -> x.f0)
+                        .transform(
+                                "findMaxIndex",
+                                Types.TUPLE(Types.INT, Types.INT),
+                                new MapPartitionFunctionWrapper<>(new FindMaxIndexFunction()));
+
+        OneHotEncoderModel model =
+                new OneHotEncoderModel()
+                        .setModelData(OneHotEncoderModelData.getModelDataTable(modelData));
+        ReadWriteUtils.updateExistingParams(model, paramMap);
+        return model;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static OneHotEncoder load(StreamExecutionEnvironment env, String path)
+            throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    /**
+     * Extract values of input columns of input data.
+     *
+     * <p>Input: rows of input data containing designated input columns
+     *
+     * <p>Output: Pairs of column index and value stored in those columns
+     */
+    private static class ExtractInputColsValueFunction
+            implements FlatMapFunction<Row, Tuple2<Integer, Integer>> {
+        private final String[] inputCols;
+
+        private ExtractInputColsValueFunction(String[] inputCols) {
+            this.inputCols = inputCols;
+        }
+
+        @Override
+        public void flatMap(Row row, Collector<Tuple2<Integer, Integer>> collector) {
+            for (int i = 0; i < inputCols.length; i++) {
+                Number number = (Number) row.getField(inputCols[i]);
+                Preconditions.checkArgument(
+                        number.intValue() == number.doubleValue(),
+                        String.format("Value %s cannot be parsed as indexed integer.", number));
+                Preconditions.checkArgument(
+                        number.intValue() >= 0, "Negative value not supported.");
+                collector.collect(new Tuple2<>(i, number.intValue()));
+            }
+        }
+    }
+
+    /** Function to find the max index value for each column. */
+    private static class FindMaxIndexFunction
+            implements MapPartitionFunction<Tuple2<Integer, Integer>, Tuple2<Integer, Integer>> {
+
+        @Override
+        public void mapPartition(
+                Iterable<Tuple2<Integer, Integer>> iterable,
+                Collector<Tuple2<Integer, Integer>> collector) {
+            Map<Integer, Integer> map = new HashMap<>();
+            for (Tuple2<Integer, Integer> value : iterable) {
+                map.put(
+                        value.f0,
+                        Math.max(map.getOrDefault(value.f0, Integer.MIN_VALUE), value.f1));
```

Comment:
Hmm, the logic here is a bit confusing to me. Should we count the number of distinct values in each group? Using the `max` here may lead the dimension of the vector to be a super large number, if the input contains a large number.

Reply:
As we discussed offline, this question is on the assumption that any categorical feature type is supported in one hot encoder. In fact the input integers should be indexed, which means if a large number appeared in input, then one hot encoder is supposed to create a sparse vector that long.

With that said, I'll add more description in Java doc stating that only indexed integer is accepted as valid input.